src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/AesCcm.Android.cs

```diff
@@ -0,0 +1,182 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.Security.Cryptography
+{
+    public sealed partial class AesCcm
+    {
+        private byte[] _key;
+
+        [MemberNotNull(nameof(_key))]
+        private void ImportKey(ReadOnlySpan<byte> key)
+        {
+            _key = key.ToArray();
+        }
+
+        private void EncryptInternal(
+            ReadOnlySpan<byte> nonce,
+            ReadOnlySpan<byte> plaintext,
+            Span<byte> ciphertext,
+            Span<byte> tag,
+            ReadOnlySpan<byte> associatedData = default)
+        {
+            // Convert key length to bits.
+            using (SafeEvpCipherCtxHandle ctx = Interop.Crypto.EvpCipherCreatePartial(GetCipher(_key.Length * 8)))
+            {
+                if (ctx.IsInvalid)
+                {
+                    throw new CryptographicException();
+                }
+
+                if (!Interop.Crypto.EvpCipherSetTagLength(ctx, tag.Length))
+                {
+                    throw new CryptographicException();
+                }
+
+                Interop.Crypto.EvpCipherSetNonceLength(ctx, nonce.Length);
+                Interop.Crypto.EvpCipherSetKeyAndIV(ctx, _key, nonce, Interop.Crypto.EvpCipherDirection.Encrypt);
+
+                if (associatedData.Length != 0)
+                {
+                    if (!Interop.Crypto.EvpCipherUpdateAAD(ctx, associatedData))
+                    {
+                        throw new CryptographicException();
+                    }
```

Comment:
FWIW, when this is the only correct implementation I've started doing

```C#
partial class Interop
{
    partial class CryptoOrWhatever
    {
        [DllImportStuff(...)]
        private static bool AndroidCryptoNative_EvpCipherUpdateAAD(...);

        internal static void EvpCipherUpdateAAD(...)
        {
            if (!AndroidCryptoNative_EvpCipherUpdateAAD(...))
            {
                throw AnAppropriateThing();
            }
        }
    }
}
```

so that it keeps "translate failure states" at the interop boundary and keeps the managed code looking more algorithmic/functional.